File: examples/combustion_example.cpp
#include "../include/state.h"
#include "../include/thermo.h"
#include "../include/combustion.h"
#include "../include/equilibrium.h"
#include "../include/humidair.h"
#include "../include/thermo_transport_data.h"
#include <iostream>
#include <iomanip>
#include <vector>

/**
 * Combustion Example: Fuel + Humid Air -> Mix -> Combust
 *
 * Demonstrates:
 * - Using set_fuel_stream_for_phi to configure fuel stream for target phi
 * - Stream mixing with enthalpy balance
 * - Complete combustion (CO2 + H2O products)
 * - WGS equilibrium (CO + H2O <-> CO2 + H2)
 * - Varying equivalence ratio from lean to stoichiometric
 */
int main()
{
    std::cout << std::fixed << std::setprecision(4);

    // =========================================================================
    // Setup: Define fuel and oxidizer streams
    // =========================================================================

    const std::size_t n_species = species_names.size();
    const std::size_t idx_CH4 = species_index_from_name("CH4");

    // Fuel stream: pure methane at 300 K (mdot will be set by set_fuel_stream_for_phi)
    Stream fuel;
    fuel.state.T = 300.0;
    fuel.state.P = 101325.0;
    fuel.state.X = std::vector<double>(n_species, 0.0);
    fuel.state.X[idx_CH4] = 1.0;

    // Oxidizer stream: humid air at 25°C, 60% RH, 10 kg/s
    Stream air;
    air.state.T = 298.15;
    air.state.P = 101325.0;
    air.state.X = humid_air_composition(air.state.T, air.state.P, 0.60);
    air.mdot = 10.0;  // kg/s (fixed)

    std::cout << "=========================================================================\n";
    std::cout << "Combustion Example: CH4 + Humid Air\n";
    std::cout << "=========================================================================\n\n";

    std::cout << "Fuel: Pure CH4 at " << fuel.state.T << " K\n";
    std::cout << "Air:  Humid air at " << air.state.T << " K, 60% RH, "
              << air.mdot << " kg/s\n\n";

    // =========================================================================
    // Sweep equivalence ratio from 0.45 to 1.0
    // =========================================================================

    std::cout << "Equivalence Ratio Sweep\n";
    std::cout << "-----------------------\n\n";

    std::cout << std::setw(6) << "phi"
              << std::setw(10) << "mdot_f"
              << std::setw(10) << "T_mix"
              << std::setw(10) << "T_ad"
              << std::setw(10) << "T_wgs"
              << std::setw(12) << "mu_ad"
              << std::setw(12) << "Pr_ad"
              << "\n";
    std::cout << std::setw(6) << "[-]"
              << std::setw(10) << "[kg/s]"
              << std::setw(10) << "[K]"
              << std::setw(10) << "[K]"
              << std::setw(10) << "[K]"
              << std::setw(12) << "[uPa.s]"
              << std::setw(12) << "[-]"
              << "\n";
    std::cout << std::string(70, '-') << "\n";

    for (double phi = 0.45; phi <= 1.01; phi += 0.05) {
        // Use set_fuel_stream_for_phi to get fuel stream with correct mdot
        Stream fuel_phi = set_fuel_stream_for_phi(phi, fuel, air);

        // Mix streams (enthalpy balance gives mixed temperature)
        Stream mixed = mix({fuel_phi, air});

        // Complete combustion (adiabatic, CO2 + H2O products)
        State burned_complete = complete_combustion(mixed.state);

        // WGS equilibrium on burned products (shifts CO + H2O <-> CO2 + H2)
        State burned_wgs = wgs_equilibrium_adiabatic(burned_complete);

        // Output results
        std::cout << std::setw(6) << phi
                  << std::setw(10) << fuel_phi.mdot
                  << std::setw(10) << mixed.state.T
                  << std::setw(10) << burned_complete.T
                  << std::setw(10) << burned_wgs.T
                  << std::setw(12) << burned_complete.mu() * 1.0e6  // convert to μPa·s
                  << std::setw(12) << burned_complete.Pr()
                  << "\n";
    }

    std::cout << "\n";

    // =========================================================================
    // Detailed output for stoichiometric case
    // =========================================================================

    std::cout << "=========================================================================\n";
    std::cout << "Detailed Results at Stoichiometric (phi = 1.0)\n";
    std::cout << "=========================================================================\n\n";

    // Use set_fuel_stream_for_phi for stoichiometric mixture
    Stream fuel_stoich = set_fuel_stream_for_phi(1.0, fuel, air);
    Stream mixed_stoich = mix({fuel_stoich, air});
    State burned_stoich = complete_combustion(mixed_stoich.state);
    State wgs_stoich = wgs_equilibrium_adiabatic(burned_stoich);

    std::cout << "Mixed Stream (before combustion):\n";
    std::cout << "  T = " << mixed_stoich.state.T << " K\n";
    std::cout << "  P = " << mixed_stoich.state.P << " Pa\n";
    std::cout << "  mdot = " << mixed_stoich.mdot << " kg/s\n";
    std::cout << "  rho = " << mixed_stoich.state.rho() << " kg/m³\n";
    std::cout << "  cp = " << mixed_stoich.state.cp() << " J/(mol·K)\n\n";

    std::cout << "Complete Combustion (CO2 + H2O only):\n";
    std::cout << "  T_ad = " << burned_stoich.T << " K\n";
    std::cout << "  rho = " << burned_stoich.rho() << " kg/m³\n";
    std::cout << "  cp = " << burned_stoich.cp() << " J/(mol·K)\n";
    std::cout << "  mu = " << burned_stoich.mu() * 1.0e6 << " μPa·s\n";
    std::cout << "  k = " << burned_stoich.k() << " W/(m·K)\n";
    std::cout << "  Pr = " << burned_stoich.Pr() << "\n";
    std::cout << "  a = " << burned_stoich.a() << " m/s\n\n";

    std::cout << "WGS Equilibrium (CO + H2O <-> CO2 + H2):\n";
    std::cout << "  T_ad = " << wgs_stoich.T << " K\n";
    std::cout << "  rho = " << wgs_stoich.rho() << " kg/m³\n";
    std::cout << "  mu = " << wgs_stoich.mu() * 1.0e6 << " μPa·s\n";
    std::cout << "  Pr = " << wgs_stoich.Pr() << "\n\n";

    // Show product composition
    std::cout << "Product Composition (WGS equilibrium, mole fractions > 0.001):\n";
    for (std::size_t i = 0; i < n_species; ++i) {
        if (wgs_stoich.X[i] > 0.001) {
            std::cout << "  " << std::setw(6) << species_names[i]
                      << ": " << std::setprecision(4) << wgs_stoich.X[i] << "\n";
        }
    }

    return 0;
}

File: python/examples/combustion_example.py
#!/usr/bin/env python3
"""Combustion example: Natural gas + humid air with combustion_equilibrium.

This example demonstrates:
- Creating fuel and air streams
- Mixing streams with enthalpy balance
- combustion_equilibrium: one-step combustion + reforming + WGS equilibrium
- Equivalence ratio sweep from lean to rich
"""

from __future__ import annotations

import numpy as np

import combaero as ca
from combaero.species import SpeciesLocator


def main() -> None:
    sp = SpeciesLocator.from_core()
    # Number of species: len(sp.names)

    # =========================================================================
    # Define fuel: Natural gas (90% CH4, 5% C2H6, 2% C3H8, 2% N2, 1% CO2)
    # =========================================================================
    X_fuel = sp.empty()
    X_fuel[sp.indices["CH4"]] = 0.90
    X_fuel[sp.indices["C2H6"]] = 0.05
    X_fuel[sp.indices["C3H8"]] = 0.02
    X_fuel[sp.indices["N2"]] = 0.02
    X_fuel[sp.indices["CO2"]] = 0.01

    fuel = ca.Stream()
    fuel.T = 300.0
    fuel.P = 101325.0
    fuel.X = X_fuel
    fuel.mdot = 1.0

    # =========================================================================
    # Define air: Humid air at 60% RH
    # =========================================================================
    X_air = np.array(ca.humid_air_composition(300.0, 101325.0, 0.6))

    air = ca.Stream()
    air.T = 300.0
    air.P = 101325.0
    air.X = X_air
    air.mdot = 10.0  # kg/s

    # =========================================================================
    # Equivalence ratio sweep
    # =========================================================================
    print("=" * 75)
    print("Combustion Example: Natural Gas + Humid Air")
    print("=" * 75)
    print(f"\nFuel: Natural gas at {fuel.T:.1f} K")
    print(
        f"      CH4: {X_fuel[sp.indices['CH4']] * 100:.0f}%, "
        f"C2H6: {X_fuel[sp.indices['C2H6']] * 100:.0f}%, "
        f"C3H8: {X_fuel[sp.indices['C3H8']] * 100:.0f}%"
    )
    print(f"Air:  Humid air at {air.T:.0f} K, 60% RH, {air.mdot:.1f} kg/s")

    print("\nEquivalence Ratio Sweep")
    print("-" * 75)
    print(f"{'phi':>6s} {'mdot_f':>10s} {'T_mix':>10s} {'T_eq':>10s} {'X_CO':>10s} {'X_H2':>10s}")
    print(f"{'[-]':>6s} {'[kg/s]':>10s} {'[K]':>10s} {'[K]':>10s} {'[%]':>10s} {'[%]':>10s}")
    print("-" * 65)

    for phi in [0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2]:
        fuel_phi = ca.set_fuel_stream_for_phi(phi, fuel, air)
        mixed = ca.mix([fuel_phi, air])
        eq = ca.combustion_equilibrium(mixed.T, mixed.X, mixed.P)

        X_CO = eq.state.X[sp.indices["CO"]] * 100
        X_H2 = eq.state.X[sp.indices["H2"]] * 100

        print(
            f"{phi:6.2f} {fuel_phi.mdot:10.4f} {mixed.T:10.2f} {eq.state.T:10.2f} {X_CO:10.4f} {X_H2:10.4f}"
        )

    # =========================================================================
    # Detailed results at stoichiometric
    # =========================================================================
    print("\n" + "=" * 75)
    print("Detailed Results at Stoichiometric (phi = 1.0)")
    print("=" * 75)

    mixed = ca.mix([ca.set_fuel_stream_for_phi(1.0, fuel, air), air])

    # One-step: combustion + reforming + WGS equilibrium
    eq = ca.combustion_equilibrium(mixed.T, mixed.X, mixed.P)

    # For comparison, also show complete combustion (before equilibrium)
    burned = ca.complete_combustion(mixed.T, mixed.X, mixed.P)

    print("\nMixed Stream (before combustion):")
    print(f"  T = {mixed.T:.2f} K")
    print(f"  P = {mixed.P:.0f} Pa")
    print(f"  mdot = {mixed.mdot:.4f} kg/s")

    print("\nComplete Combustion (CO2 + H2O only):")
    print(f"  T_ad = {burned.T:.2f} K")

    print("\nCombustion + Equilibrium (one-step):")
    print(f"  T_eq = {eq.state.T:.2f} K")

    print("\nProduct Composition (mole fractions > 0.1%):")
    for i, name in enumerate(sp.names):
        if eq.state.X[i] > 0.001:
            print(f"  {name:>8s}: {eq.state.X[i] * 100:6.2f}%")

    # =========================================================================
    # Rich case (phi = 1.2) - Reforming converts hydrocarbons to CO + H2
    # =========================================================================
    print("\n" + "=" * 75)
    print("Rich Case (phi = 1.2) - Reforming + WGS Equilibrium")
    print("=" * 75)

    mixed = ca.mix([ca.set_fuel_stream_for_phi(1.2, fuel, air), air])

    # One-step: combustion + reforming + WGS equilibrium
    eq = ca.combustion_equilibrium(mixed.T, mixed.X, mixed.P)

    # For comparison, also show complete combustion (before equilibrium)
    burned = ca.complete_combustion(mixed.T, mixed.X, mixed.P)

    print("\nComplete Combustion (before equilibrium):")
    print(f"  T_ad = {burned.T:.2f} K")
    print("  Unburned hydrocarbons:")
    for hc in ["CH4", "C2H6", "C3H8"]:
        if burned.X[sp.indices[hc]] > 1e-6:
            print(f"    {hc}: {burned.X[sp.indices[hc]] * 100:.4f}%")

    print("\nCombustion + Equilibrium (one-step, hydrocarbons reformed to CO + H2):")
    print(f"  T_eq = {eq.state.T:.2f} K (dropped {burned.T - eq.state.T:.1f} K)")
    print("  Product composition:")
    for i, name in enumerate(sp.names):
        if eq.state.X[i] > 0.001:
            print(f"    {name:>8s}: {eq.state.X[i] * 100:6.2f}%")


if __name__ == "__main__":
    main()
